### PR TITLE
Add coordinated shutdown support

### PR DESCRIPTION
**Pull request by ennru**
Add helpers to add Akka HTTP server bindings to Akka Coordinated shutdown.

References #1210

### CODE REVIEW

File: docs/src/main/paradox/server-side/graceful-termination.md [hunk 1]
```diff
@@ -1,5 +1,33 @@
...
 
+## Akka Coordinated Shutdown
+
+Coordinated shutdown is Akka's managed way of shutting down multiple modules / sub-systems (persistence, cluster, http etc)
+in a predictable and ordered fashion. For example, in a typical Akka application you will want to stop accepting new HTTP connections, and then shut down the cluster etc.
+
+Akka HTTP shutdown consists mainly of three steps:
+
+1. stop accepting new connections (@apidoc[ServerBinding.unbind](ServerBinding))
+1. close open connections (@apidoc[ServerBinding.terminat](ServerBinding))
```
Comment (jrudolph) [inline docs/src/main/paradox/server-side/graceful-termination.md:11]:
```suggestion
1. close open connections (@apidoc[ServerBinding.terminate](ServerBinding))
```

File: docs/src/test/scala/docs/http/scaladsl/ServerShutdownExampleSpec.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+import scala.concurrent.duration._
+import scala.concurrent._
+import scala.io.StdIn
+
+class ServerShutdownExampleSpec extends AnyWordSpec with Matchers
+  with CompileOnlySpec {
+
+  // #suggested
+  def addToCoordinatedShutdown(terminationGracePeriod: FiniteDuration)(binding: ServerBinding)(implicit system: ClassicActorSystemProvider): ServerBinding = {
```
Comment (jrudolph) [inline docs/src/test/scala/docs/http/scaladsl/ServerShutdownExampleSpec.scala:26]: Wouldn't people copy and paste this exact snapshot everywhere? Could we either support it as a configuration option or at least just provide that exact method somewhere in the Http extension?

**Comment by ennru** [comment]
I wasn't sure if we'd want to add this as code, nor where.

This updated variant adds it directly to `ServerBinding`. WDYT?

File: docs/src/test/scala/docs/http/scaladsl/server/ServerShutdownExampleSpec.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    Await.ready(
+      bindingFuture.flatMap(_ => waitForShutdownSignal(system)), // chaining both futures to fail fast
+      Duration.Inf) // It's waiting forever because maybe there is never a shutdown signal
+
+    // #shutdown
+    case object UserInitiatedShutdown extends CoordinatedShutdown.Reason
+
+    CoordinatedShutdown(system).run(UserInitiatedShutdown)
```
Comment (jrudolph) [inline docs/src/test/scala/docs/http/scaladsl/server/ServerShutdownExampleSpec.scala:55]: Is that necessary to trigger coordinated shutdown? What happens if you just use `ActorSystem.terminate`?

Comment (ennru) [reply docs/src/test/scala/docs/http/scaladsl/server/ServerShutdownExampleSpec.scala:55]: It can be done this way to indicate the shutdown reason, `ActorSystem.terminate` will use `ActorSystemTerminateReason`.
But this should maybe go into the main Akka docs instead.

Comment (ennru) [reply docs/src/test/scala/docs/http/scaladsl/server/ServerShutdownExampleSpec.scala:42]: I added `ActorSystem.terminate` as an alternative.
This PR goes along with https://github.com/akka/akka/pull/29057

File: docs/src/main/paradox/server-side/graceful-termination.md [hunk 4]
```diff
@@ -1,5 +1,33 @@
...
+## Akka Coordinated Shutdown
+
+Coordinated shutdown is Akka's managed way of shutting down multiple modules / sub-systems (persistence, cluster, http etc)
+in a predictable and ordered fashion. For example, in a typical Akka application you will want to stop accepting new HTTP connections, and then shut down the cluster etc.
+
+Akka HTTP shutdown consists mainly of three steps:
+
+1. stop accepting new connections (@scala[@scaladoc[ServerBinding.unbind](akka.http.scaladsl.Http.ServerBinding)]@java[@javadoc[ServerBinding.unbind](akka.http.javadsl.ServerBinding)])
+1. close open connections (@scala[@scaladoc[ServerBinding.terminate](akka.http.scaladsl.Http.ServerBinding)]@java[@javadoc[ServerBinding.terminate](akka.http.javadsl.ServerBinding)])
+1. shut down connection pools (@scala[@scaladoc[Http.shutdownAllConnectionPools](akka.http.scaladsl.HttpExt)]@java[@javadoc[Http.shutdownAllConnectionPools](akka.http.javadsl.Http)])
```
Comment (hseeberger) [inline docs/src/main/paradox/server-side/graceful-termination.md:12]: Why is this necessary at all after unbinding and terminating?

Comment (raboof) [reply docs/src/main/paradox/server-side/graceful-termination.md:12]: the first 2 items are about the server side, `shutdownAllConnectionPools` is about the client side.

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 5]
```diff
@@ -999,6 +999,25 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
...
+    def addToCoordinatedShutdown(hardTerminationDeadline: FiniteDuration)(implicit system: ClassicActorSystemProvider): ServerBinding = {
+      val shutdown = CoordinatedShutdown(system)
+      shutdown.addTask(CoordinatedShutdown.PhaseServiceUnbind, s"http-unbind-${localAddress}") { () =>
+        unbind()
+      }
+      shutdown.addTask(CoordinatedShutdown.PhaseServiceRequestsDone, s"http-terminate-${localAddress}") { () =>
+        terminate(hardTerminationDeadline).map(_ => Done)(ExecutionContexts.sameThreadExecutionContext)
+      }
+      shutdown.addTask(CoordinatedShutdown.PhaseServiceStop, s"http-shutdown-${localAddress}") { () =>
+        Http()(system).shutdownAllConnectionPools().map(_ => Done)(ExecutionContexts.sameThreadExecutionContext)
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:1017]: Since this is about the client side, it seems weird that `addToCoordinatedShutdown` on the `ServerBinding` would cause those to be closed.

Shutting down the client connection pools seems less important anyway, perhaps we should just remove this task here and let the pools shut down along with the rest of the actor system? Or is there a scenario where that is messy?

Comment (ennru) [reply akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:1017]: Good point.
I'll come up with separate information for the client. There, it is more up to the application to stop ongoing requests before they shut down the pools and the actor system.

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 6]
```diff
@@ -744,6 +744,17 @@ class HttpExt private[http] (private val config: Config)(implicit val system: Ex
     shutdownCompletedPromise.future.map(_ => ())(system.dispatcher)
   }
 
+  /**
+   * Adds the connection pools to the actor system's coordinated shutdown, so that [[shutdownAllConnectionPools]] gets
+   * called before the actor system finally terminates.
+   */
+  def addClientPoolsToCoordinatedShutdown(): Unit = {
```
Comment (ennru) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:751]: I added this, but for client pools we might as well add it to coordinated shutdown automatically.

File: akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala [hunk 7]
```diff
@@ -112,6 +111,18 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
     delegate.whenTerminated
       .map(_.asInstanceOf[HttpTerminated])(ExecutionContexts.sameThreadExecutionContext)
       .toJava
+
+  /**
+   * Adds this `ServerBinding` to the actor system's coordinated shutdown, so that [[unbind]] and [[terminate]] get
+   * called appropriately before the connection pools are shut down.
```
Comment (hseeberger) [inline akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala:117]: "before the connection pools are shut down"? I guess this should be rephrased "before the actor system is shut down (or terminated)".

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala (hunk 5 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: > Shutting down the client connection pools seems less important anyway, perhaps we should just remove this task here and let the pools shut down along with the rest of the actor system? Or is there a scenario where that is messy?

It's noisy because the actors are running and pulled out under them. But maybe the client side shutting down could be automatically added to coordinated shutdown. But that can be a different PR.

File: docs/src/main/paradox/server-side/graceful-termination.md [hunk 8]
```diff
@@ -1,5 +1,32 @@
 # Graceful termination
 
+## Akka Coordinated Shutdown
+
+@extref[Coordinated shutdown](akka-docs:coordinated-shutdown.html) is Akka's managed way of shutting down multiple modules / sub-systems (persistence, cluster, http etc)
+in a predictable and ordered fashion. For example, in a typical Akka application you will want to stop accepting new HTTP connections, and then shut down the cluster etc.
+
+The recommended Akka HTTP server shutdown consists of two steps:
+
+1. stop accepting new connections (@scala[@scaladoc[ServerBinding.unbind](akka.http.scaladsl.Http.ServerBinding)]@java[@javadoc[ServerBinding.unbind](akka.http.javadsl.ServerBinding)])
```
Comment (jrudolph) [inline docs/src/main/paradox/server-side/graceful-termination.md:10]:
```suggestion
1. stop accepting new connections (@scala[@scaladoc[ServerBinding.unbind](akka.http.scaladsl.Http.ServerBinding)]@java[@javadoc[ServerBinding.unbind](akka.http.javadsl.ServerBinding)])
1. try to finish handling of ongoing requests until the `hardTerminationDeadline` hits (see below for details)
```

File: docs/src/main/paradox/server-side/graceful-termination.md [hunk 9]
```diff
@@ -1,5 +1,32 @@
 # Graceful termination
 
+## Akka Coordinated Shutdown
+
+@extref[Coordinated shutdown](akka-docs:coordinated-shutdown.html) is Akka's managed way of shutting down multiple modules / sub-systems (persistence, cluster, http etc)
+in a predictable and ordered fashion. For example, in a typical Akka application you will want to stop accepting new HTTP connections, and then shut down the cluster etc.
+
+The recommended Akka HTTP server shutdown consists of two steps:
```
Comment (jrudolph) [inline docs/src/main/paradox/server-side/graceful-termination.md:8]:
```suggestion
The recommended Akka HTTP server shutdown consists of three steps:
```

File: docs/src/test/scala/docs/http/scaladsl/server/ServerShutdownExampleSpec.scala (hunk 3 above)
Comment (jrudolph) [reply docs/src/test/scala/docs/http/scaladsl/server/ServerShutdownExampleSpec.scala:42]: Thanks!

**Comment by jroper** [comment]
Nice.
